spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,483 @@
+- `delayPeriodBlocks`: The delay period MAY be modified by the upgrade protocol. The counterparty MUST accept the new proposed value or return an error during the upgrade handshake.
+
+MUST NOT BE MODIFIED:
+- `counterpartyConnectionIdentifier`: The counterparty connection identifier CAN NOT be modified by the upgrade protocol.
+- `clientIdentifier`: The client identifier CAN NOT be modified by the upgrade protocol
+- `counterpartyClientIdentifier`: The counterparty client identifier CAN NOT be modified by the upgrade protocol
+
+NOTE: If the upgrade adds any fields to the `ConnectionEnd` these are by default modifiable, and can be arbitrarily chosen by an Actor (e.g. chain governance) which has permission to initiate the upgrade.
+
+```typescript
+interface UpgradeTimeout {
+    timeoutHeight: Height
+    timeoutTimestamp: uint64
+}
+```
+
+- `timeoutHeight`: Timeout height indicates the height at which the counterparty must no longer proceed with the upgrade handshake. The chains will then preserve their original connection and the upgrade handshake is aborted.
+- `timeoutTimestamp`: Timeout timestamp indicates the time on the counterparty at which the counterparty must no longer proceed with the upgrade handshake. The chains will then preserve their original connection and the upgrade handshake is aborted.
+
+At least one of the timeoutHeight or timeoutTimestamp MUST be non-zero.
+
+### Store Paths
+
+#### Restore Connection Path
+
+The chain must store the previous connection end so that it may restore it if the upgrade handshake fails. This may be stored in the private store.
+
+```typescript
+function restorePath(id: Identifier): Path {
+    return "connections/{id}/restore"
+}
+```
+
+#### UpgradeError Path
+
+The upgrade error path is a public path that can signal an error of the upgrade to the counterparty. It does not store anything in the successful case, but it will store a sentinel abort value in the case that a chain does not accept the proposed upgrade.
+
+```typescript
+function errorPath(id: Identifier): Path {
+    return "connections/{id}/upgradeError"
+
+}
+```
+
+The UpgradeError MUST have an associated verification function added to the connection and client interfaces so that a counterparty may verify that chain has stored an error in the UpgradeError path.
+
+```typescript
+// Connection VerifyConnectionUpgradeError method
+function verifyConnectionUpgradeError(
+  connection: ConnectionEnd,
+  height: Height,
+  proof: CommitmentProof,
+  upgradeErrorReceipt: []byte, 
+) {
+    client = queryClient(connection.clientIdentifier)
+    client.verifyUpgradeError(height, connection.counterpartyPrefix, proof, connection.counterpartyConnectionIdentifier, upgradeErrorReceipt)
+}
+```
+
+```typescript
+// Client VerifyUpgradeError
+function verifyConnectionUpgradeError(
```

Comment:
```suggestion
function VerifyUpgradeError(
```

Reply:
I've added Connection here to the name because there are identical functions for channel upgradability.